### ISSUE TITLE
Regression in 2.2.1: evaluating a string starting with a pound sign errors out

### ISSUE THREAD

**Issue by moritz** [ansible/ansible#21012]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
core (?)

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION

Standard

##### OS / ENVIRONMENT
N/A

##### SUMMARY

Using a string starting with a `#` in a `when:` clause dies with `The conditional check 'smartd_conf' failed. The error was: unexpected char u'#' at 6\n  line 1\n\nThe error appears to have been in '/home/moritz/foo.yaml': line 8, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - command: echo test\n      ^ here\n`

##### STEPS TO REPRODUCE

Execute the following playbook on Ansible 2.2.1:

```yaml
---
- hosts: localhost
  connection: local
  vars:
    smartd_conf: |
      # mercury has three SCSI devices: /dev/sda is the Dell PERC H200 RAID adapter
  tasks:
    - command: echo test
      when: smartd_conf
```

##### EXPECTED RESULTS
The task with the condition should run, because a non-empty string evaluated as a boolean should be `True`. That's what it did in Ansible 2.0, and supposedly also on 2.2.0

##### ACTUAL RESULTS

```
TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'smartd_conf' failed. The error was: unexpected char u'#' at 6\n  line 1\n\nThe error appears to have been in '/home/moritz/foo.yaml': line 8, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - command: echo test\n      ^ here\n"}
	to retry, use: --limit @/home/moritz/foo.retry

```

**Comment by jeremyn** [comment]
A '|' causes a similar problem. Playbook:

```yaml
---
- hosts: localhost
  connection: local
  vars:
    smartd_conf: |
      | mercury has three SCSI devices: /dev/sda is the Dell PERC H200 RAID adapter  
  tasks:
    - command: echo test
      when: smartd_conf
```

Output:

```
TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'smartd_conf' failed. The error was: unexpected '|'\n  line 1\n\nThe error appears to have been in '/vagrant/playbook.yml': line 8, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - command: echo test\n      ^ here\n"}
```

My use case is writing an SSH `known_hosts` file only `when` it has non-empty content, and the content started with a `|`.

That said, @moritz: I don't think non-empty strings normally evaluate as True. For example with:

```yaml
---
- hosts: localhost
  connection: local
  vars:
    smartd_conf: mercury
  tasks:
    - command: echo test
      when: smartd_conf
```

I get:

```
TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'smartd_conf' failed. The error was: error while evaluating conditional (smartd_conf): 'mercury' is undefined\n\nThe error appears to have been in '/vagrant/playbook.yml': line 7, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - command: echo test\n      ^ here\n"}
```

You need to be more explicit like with `when: smartd_conf != ""`, or "trick" it with
`when: '"{{ smartd_conf }}"'`, which then causes the `command` to activate or not as appropriate.

I also looked into using a [`bool`](https://github.com/ansible/ansible/blob/d0d1158c5ef7c1f822e622bebd80eb9471285b0b/lib/ansible/plugins/filter/core.py#L109-L118) filter, but that only evaluates as True with specific non-empty values.

**Comment by samdoran** [comment]
Using `is defined`, `!= ''`, or `| default(False)` in the conditional is the proper way to test this.

**Comment by mgedmin** [comment]
The original example looks like it came from my playbooks.  I don't remember what happened here -- most likely I complained on IRC and somebody else filed a bug?

My goal was to run a task when a variable (let's call it `x` for simplicity) had a non-blank value, where non-blank means "is not an empty string or a null value".  In Python I would write `if x:`, which is why I originally attempted `when: x` in Ansible.

(A reminder that null is what you get when you define a variable like this:

```yaml
---
x:
```

From the proposed alternatives:

- `x is defined` runs when x is "" or null.

- `x != ''` runs when x is null

- `x|default(False)` does exactly what I want it to do!

- `not not x` is the workaround I went with back in January 2017, and I'm happy I can replace it with something less obscure.

Thank you
